### ISSUE TITLE
如何构建特征工程多重线性过滤特征

### ISSUE THREAD

**Issue by laitianan** [laitianan/DeepFM#1]
import pandas as pd
import numpy as np

from scipy.stats import pearsonr, pointbiserialr, f_oneway, chi2_contingency
from sklearn.feature_selection import SelectKBest, f_regression, mutual_info_regression
from sklearn.preprocessing import LabelEncoder, StandardScaler
import warnings
warnings.filterwarnings('ignore')

class SmartCorrelationFeatureSelector:
    """
    智能相关性特征筛选器（支持数值型和类别型特征）
    """
    
    def __init__(self, target_col, correlation_threshold=0.95, 
                 target_correlation_threshold=0.01, encode_categorical=True):
        """
        初始化
        
        参数:
        target_col: 目标变量列名
        correlation_threshold: 特征间相关性阈值
        target_correlation_threshold: 与目标变量相关性阈值
        encode_categorical: 是否对类别特征进行编码
        """
        self.target_col = target_col
        self.correlation_threshold = correlation_threshold
        self.target_correlation_threshold = target_correlation_threshold
        self.encode_categorical = encode_categorical
        
        self.feature_correlation_matrix_ = None
        self.target_correlations_ = None
        self.selected_features_ = None
        self.feature_types_ = None
        self.encoders_ = {}
        
    def _detect_feature_types(self, df):
        """
        检测特征类型
        """
        X = df.drop(columns=[self.target_col], errors='ignore')
        
        feature_types = {}
        for col in X.columns:
            # 检查数据类型和唯一值数量
            dtype = str(X[col].dtype)
            n_unique = X[col].nunique()
            sample_non_na = X[col].dropna().iloc[0] if not X[col].dropna().empty else None
            
            if (dtype in ['object', 'category'] or 
                (n_unique <= 20 and isinstance(sample_non_na, (str, bool))) or
                (n_unique <= 10 and X[col].notna().sum() > 0)):
                feature_types[col] = 'categorical'
            else:
                feature_types[col] = 'numerical'
        
        return feature_types
    
    def _encode_categorical_features(self, X):
        """
        对类别特征进行编码
        """
        X_encoded = X.copy()
        self.encoders_ = {}
        
        for col in X.columns:
            if self.feature_types_.get(col) == 'categorical':
                # 使用LabelEncoder进行编码
                le = LabelEncoder()
                # 处理缺失值
                non_na_values = X[col].dropna()
                if len(non_na_values) > 0:
                    try:
                        # 只对非空值进行拟合转换
                        le.fit(non_na_values.astype(str))
                        X_encoded[col] = X[col].astype(str)
                        X_encoded.loc[X[col].notna(), col] = le.transform(X[col].dropna().astype(str))
                        X_encoded[col] = pd.to_numeric(X_encoded[col], errors='coerce')
                        self.encoders_[col] = le
                    except Exception as e:
                        print(f"编码特征 {col} 时出错: {e}")
                        # 如果编码失败，使用简单数值编码
                        X_encoded[col] = pd.factorize(X[col])[0]
                else:
                    # 如果全是空值，设为0
                    X_encoded[col] = 0
        
        return X_encoded
    
    def _calculate_correlation_mixed(self, col1, col2, X, feature_types):
        """
        计算混合类型特征间的相关性
        """
        # 处理缺失值
        valid_idx = X[col1].notna() & X[col2].notna()
        if valid_idx.sum() < 10:  # 至少需要10个有效样本
            return 0
        
        try:
            if feature_types.get(col1) == 'numerical' and feature_types.get(col2) == 'numerical':
                # 数值型 vs 数值型：使用Pearson相关系数
                corr, _ = pearsonr(X.loc[valid_idx, col1], X.loc[valid_idx, col2])
                return corr
                    
            elif feature_types.get(col1) == 'categorical' and feature_types.get(col2) == 'categorical':
                # 类别型 vs 类别型：使用Cramér's V
                contingency_table = pd.crosstab(X[col1], X[col2])
                if contingency_table.size == 0:
                    return 0
                    
                chi2, p, dof, expected = chi2_contingency(contingency_table)
                n = contingency_table.sum().sum()
                phi2 = chi2 / n
                r, k = contingency_table.shape
                phi2corr = max(0, phi2 - ((k-1)*(r-1))/(n-1))
                rcorr = r - ((r-1)**2)/(n-1)
                kcorr = k - ((k-1)**2)/(n-1)
                return np.sqrt(phi2corr / min((kcorr-1), (rcorr-1)))
                    
            else:
                # 数值型 vs 类别型：使用相关比（eta squared）
                if feature_types.get(col1) == 'categorical':
                    categorical_col, numerical_col = col1, col2
                else:
                    categorical_col, numerical_col = col2, col1
                
                groups = [group[numerical_col].dropna() for name, group in 
                         X.groupby(categorical_col) if len(group) > 0]
                
                if len(groups) < 2:
                    return 0
                
                f_stat, _ = f_oneway(*groups)
                eta_squared = f_stat / (f_stat + (valid_idx.sum() - len(groups)))
                return np.sqrt(eta_squared)
                
        except Exception as e:
            print(f"计算特征 {col1} 和 {col2} 相关性时出错: {e}")
            return 0
    
    def _calculate_target_correlation_mixed(self, feature_col, target_col, X, y, feature_type, target_type):
        """
        计算特征与目标变量的相关性（混合类型）
        """
        valid_idx = X[feature_col].notna() & y.notna()
        if valid_idx.sum() < 10:
            return {'correlation': 0, 'p_value': 1}
        
        try:
            if target_type == 'numerical':
                # 目标变量是数值型
                if feature_type == 'numerical':
                    # 数值特征 vs 数值目标：Pearson
                    corr, p_value = pearsonr(X.loc[valid_idx, feature_col], y[valid_idx])
                    return {'correlation': corr, 'p_value': p_value}
                else:
                    # 类别特征 vs 数值目标：相关比
                    groups = [y[group.index] for name, group in 
                             X.groupby(feature_col) if len(group) > 0]
                    
                    if len(groups) < 2:
                        return {'correlation': 0, 'p_value': 1}
                    
                    f_stat, p_value = f_oneway(*groups)
                    eta_squared = f_stat / (f_stat + (valid_idx.sum() - len(groups)))
                    correlation = np.sqrt(eta_squared)
                    return {'correlation': correlation, 'p_value': p_value}
                    
            else:
                # 目标变量是类别型
                if feature_type == 'numerical':
                    # 数值特征 vs 类别目标：点二列相关
                    y_encoded = LabelEncoder().fit_transform(y[valid_idx].astype(str))
                    corr, p_value = pointbiserialr(X.loc[valid_idx, feature_col], y_encoded)
                    return {'correlation': corr, 'p_value': p_value}
                else:
                    # 类别特征 vs 类别目标：Cramér's V
                    contingency_table = pd.crosstab(X[feature_col], y)
                    if contingency_table.size == 0:
                        return {'correlation': 0, 'p_value': 1}
                        
                    chi2, p_value, dof, expected = chi2_contingency(contingency_table)
                    n = contingency_table.sum().sum()
                    phi2 = chi2 / n
                    r, k = contingency_table.shape
                    phi2corr = max(0, phi2 - ((k-1)*(r-1))/(n-1))
                    rcorr = r - ((r-1)**2)/(n-1)
                    kcorr = k - ((k-1)**2)/(n-1)
                    correlation = np.sqrt(phi2corr / min((kcorr-1), (rcorr-1)))
                    return {'correlation': correlation, 'p_value': p_value}
                    
        except Exception as e:
            print(f"计算特征 {feature_col} 与目标变量相关性时出错: {e}")
            return {'correlation': 0, 'p_value': 1}
    
    def calculate_correlations(self, df):
        """
        计算相关性矩阵（支持混合类型特征）
        """
        # 检测特征类型
        self.feature_types_ = self._detect_feature_types(df)
        
        # 分离特征和目标变量
        X = df.drop(columns=[self.target_col], errors='ignore')
        y = df[self.target_col] if self.target_col in df.columns else None
        
        # 检测目标变量类型
        if y is not None:
            target_type = 'categorical' if y.dtype == 'object' or y.nunique() <= 10 else 'numerical'
        else:
            target_type = None
        
        print(f"检测到特征类型: {self.feature_types_}")
        
        # 如果需要编码且存在类别特征，进行编码
        if self.encode_categorical and any(typ == 'categorical' for typ in self.feature_types_.values()):
            X_encoded = self._encode_categorical_features(X)
            print("已完成类别特征编码")
        else:
            X_encoded = X.copy()
        
        # 计算特征间相关性矩阵
        corr_matrix = pd.DataFrame(index=X.columns, columns=X.columns)
        
        print("计算特征间相关性矩阵...")
        for i, col1 in enumerate(X.columns):
            for j, col2 in enumerate(X.columns):
                if i == j:
                    corr_matrix.loc[col1, col2] = 1.0
                elif i < j:
                    corr = self._calculate_correlation_mixed(col1, col2, X_encoded, self.feature_types_)
                    corr_matrix.loc[col1, col2] = corr
                    corr_matrix.loc[col2, col1] = corr
        
        self.feature_correlation_matrix_ = corr_matrix.astype(float)
        
        # 计算与目标变量的相关性
        if y is not None:
            print("计算特征与目标变量的相关性...")
            target_correlations = {}
            for col in X.columns:
                result = self._calculate_target_correlation_mixed(
                    col, self.target_col, X_encoded, y, 
                    self.feature_types_.get(col, 'numerical'), target_type
                )
                target_correlations[col] = result
            
            self.target_correlations_ = pd.DataFrame(target_correlations).T
            self.target_correlations_ = self.target_correlations_.sort_values(
                'correlation', key=abs, ascending=False
            )
        
        return self
    
    def find_highly_correlated_features(self):
        """
        找出高度相关的特征对
        """
        if self.feature_correlation_matrix_ is None:
            raise ValueError("请先调用 calculate_correlations 方法")
        
        corr_matrix = self.feature_correlation_matrix_.copy()
        np.fill_diagonal(corr_matrix.values, 0)
        
        high_corr_pairs = []
        
        for i in range(len(corr_matrix.columns)):
            for j in range(i+1, len(corr_matrix.columns)):
                corr_value = abs(corr_matrix.iloc[i, j])
                if corr_value > self.correlation_threshold:
                    high_corr_pairs.append({
                        'feature1': corr_matrix.columns[i],
                        'feature2': corr_matrix.columns[j],
                        'correlation': corr_value,
                        'type1': self.feature_types_.get(corr_matrix.columns[i], 'unknown'),
                        'type2': self.feature_types_.get(corr_matrix.columns[j], 'unknown')
                    })
        
        high_corr_pairs.sort(key=lambda x: x['correlation'], reverse=True)
        return high_corr_pairs
    
    def select_features(self, df, method='hybrid'):
        """
        特征选择主函数
        """
        # 计算相关性
        self.calculate_correlations(df)
        
        X = df.drop(columns=[self.target_col])
        y = df[self.target_col]
        
        # 基于特征类型选择策略
        numerical_features = [f for f, t in self.feature_types_.items() if t == 'numerical']
        categorical_features = [f for f, t in self.feature_types_.items() if t == 'categorical']
        
        print(f"\n特征类型统计:")
        print(f"  数值型特征: {len(numerical_features)}个")
        print(f"  类别型特征: {len(categorical_features)}个")
        
        if method == 'hybrid':
            selected_features = self._select_hybrid(X, y)
        else:
            selected_features = self._select_simple(X, y)
        
        self.selected_features_ = selected_features
        return selected_features
    
    def _select_hybrid(self, X, y):
        """混合选择方法"""
        selected_features = []
        
        # 对数值型特征：使用相关性筛选
        numerical_features = [f for f, t in self.feature_types_.items() if t == 'numerical']
        if numerical_features:
            num_correlations = self.target_correlations_.loc[numerical_features]
            significant_numerical = num_correlations[
                (abs(num_correlations['correlation']) > self.target_correlation_threshold)
            ].index.tolist()
            selected_features.extend(significant_numerical)
            print(f"选择数值型特征: {len(significant_numerical)}个")
        
        # 对类别型特征：基于相关性筛选
        categorical_features = [f for f, t in self.feature_types_.items() if t == 'categorical']
        if categorical_features:
            cat_correlations = self.target_correlations_.loc[categorical_features]
            significant_categorical = cat_correlations[
                (abs(cat_correlations['correlation']) > self.target_correlation_threshold)
            ].index.tolist()
            selected_features.extend(significant_categorical)
            print(f"选择类别型特征: {len(significant_categorical)}个")
        
        # 移除高相关性特征
        final_features = self._remove_high_correlations(X[selected_features])
        print(f"移除高相关性特征后剩余: {len(final_features)}个")
        
        return final_features
    
    def _remove_high_correlations(self, X):
        """移除高相关性特征"""
        if len(X.columns) <= 1:
            return X.columns.tolist()
        
        features_to_keep = set(X.columns)
        features_removed = []
        
        for i in range(len(X.columns)):
            feature_i = X.columns[i]
            if feature_i not in features_to_keep:
                continue
                
            for j in range(i+1, len(X.columns)):
                feature_j = X.columns[j]
                if feature_j not in features_to_keep:
                    continue
                    
                try:
                    corr_value = abs(self.feature_correlation_matrix_.loc[feature_i, feature_j])
                    if corr_value > self.correlation_threshold:
                        # 保留与目标变量相关性更高的特征
                        if self.target_correlations_ is not None:
                            corr_i = abs(self.target_correlations_.loc[feature_i, 'correlation'])
                            corr_j = abs(self.target_correlations_.loc[feature_j, 'correlation'])
                            
                            if corr_i >= corr_j:
                                features_to_keep.discard(feature_j)
                                features_removed.append(feature_j)
                            else:
                                features_to_keep.discard(feature_i)
                                features_removed.append(feature_i)
                                break  # 跳出内层循环，因为feature_i已被移除
                except:
                    continue
        
        if features_removed:
            print(f"移除的高相关性特征: {features_removed}")
        
        return list(features_to_keep)
    
    def get_feature_type_summary(self):
        """获取特征类型摘要"""
        if self.feature_types_ is None:
            return {}
        
        type_count = {}
        for feature_type in self.feature_types_.values():
            type_count[feature_type] = type_count.get(feature_type, 0) + 1
        
        return type_count
    
    def visualize_correlations(self, figsize=(15, 12)):
        """可视化相关性矩阵"""
        import seaborn as sns
        import matplotlib.pyplot as plt
        if self.feature_correlation_matrix_ is None:
            raise ValueError("请先调用 calculate_correlations 方法")
        
        fig, axes = plt.subplots(2, 2, figsize=figsize)
        
        # 1. 特征间相关性热力图
        plt.subplot(2, 2, 1)
        mask = np.triu(np.ones_like(self.feature_correlation_matrix_, dtype=bool))
        sns.heatmap(self.feature_correlation_matrix_, mask=mask, cmap='coolwarm', 
                   center=0, annot=False, fmt='.2f')
        plt.title('特征间相关性热力图')
        
        # 2. 与目标变量的相关性条形图
        if self.target_correlations_ is not None:
            plt.subplot(2, 2, 2)
            top_features = self.target_correlations_.head(20)
            colors = ['red' if x < 0 else 'blue' for x in top_features['correlation']]
            plt.barh(range(len(top_features)), top_features['correlation'], color=colors)
            plt.yticks(range(len(top_features)), top_features.index)
            plt.xlabel('与目标变量的相关性')
            plt.title('特征与目标变量相关性排名(Top 20)')
        
        plt.tight_layout()
        plt.show()

def automated_smart_feature_selection(df, target_col, correlation_threshold=0.9, 
                                    target_correlation_threshold=0.05):
    """
    自动化智能特征选择流程（支持类别特征）
    """
    print("=" * 60)
    print("智能特征筛选开始（支持类别特征）")
    print("=" * 60)
    
    # 初始化选择器
    selector = SmartCorrelationFeatureSelector(
        target_col=target_col,
        correlation_threshold=correlation_threshold,
        target_correlation_threshold=target_correlation_threshold,
        encode_categorical=True
    )
    
    # 执行特征选择
    selected_features = selector.select_features(df)
    
    # 显示结果
    feature_types = selector.get_feature_type_summary()
    print(f"\n特征类型分布: {feature_types}")
    
    # 显示高相关性特征对
    high_corr_pairs = selector.find_highly_correlated_features()
    print(f"\n高相关性特征对数量(> {correlation_threshold}): {len(high_corr_pairs)}")
    
    if high_corr_pairs:
        print("前5对高相关性特征:")
        for i, pair in enumerate(high_corr_pairs[:5]):
            print(f"  {i+1}. {pair['feature1']}({pair['type1']}) vs {pair['feature2']}({pair['type2']}): {pair['correlation']:.4f}")
    
    print(f"\n特征选择完成!")
    print(f"原始特征数量: {len(df.columns) - 1}")
    print(f"筛选后特征数量: {len(selected_features)}")
    print(f"特征减少比例: {(len(df.columns) - 1 - len(selected_features)) / (len(df.columns) - 1):.2%}")
    
    # 创建筛选后的数据集
    final_features = selected_features + [target_col]
    df_selected = df[final_features].copy()
    
    result = {
        'selected_df': df_selected,
        'selected_features': selected_features,
        'selector': selector,
        'feature_types': feature_types,
        'high_corr_pairs': high_corr_pairs
    }
    
    return result

# 使用示例
if __name__ == "__main__":
    # 读取数据
    df = pd.read_csv('./data/smart_selected_df.csv')
    
    # 自动检测目标变量
    if 'target' in df.columns:
        target_column = 'target'
    elif 'label' in df.columns:
        target_column = 'label'
    else:
        # 选择最后一列作为目标变量
        target_column = 'log_time_to_delivery'
        print(f"自动选择最后一列作为目标变量: {target_column}")
    
    # 检查数据中的类别特征
    categorical_cols = df.select_dtypes(include=['object', 'category']).columns
    numerical_cols = df.select_dtypes(include=[np.number]).columns
    
    print(f"数据形状: {df.shape}")
    print(f"类别特征: {len(categorical_cols)}个")
    print(f"数值特征: {len(numerical_cols)}个")
    
    # 执行智能特征选择
    result = automated_smart_feature_selection(
        df=df,
        target_col=target_column,
        correlation_threshold=0.9,
        target_correlation_threshold=0.05
    )
    
    # 保存结果
    result['selected_df'].to_csv('./data/smart_selected_df_smart_filtered.csv', index=False)
    print(f"\n筛选后的数据已保存为: ./data/smart_selected_df_smart_filtered.csv")
    
    # 可选：可视化相关性
    # result['selector'].visualize_correlations()
